### ISSUE TITLE
Cannot create new user as Admin

### ISSUE THREAD

**Issue by mk4111** [emfoundation/ce100-app#475]
When we try to create/submit the details of a new user, the app breaks with a 500 error:

![when-adding-new-user](https://cloud.githubusercontent.com/assets/13470325/21190399/85f8e302-c219-11e6-8e56-b289a75667d3.png)

All fields have been filled in, but nothing is saved in the database, so the organisation remains primary-user-less

**Comment by mk4111** [comment]
This error only seems to happen when the inputted _email_ already exists in the database, because of the postgres `UNIQUE` constraint.

Need to find a better way to fail/let the user know the email already exists.

**Comment by mk4111** [comment]
@JMurphyWeb and @SimonLab what are your thoughts on this?

**Comment by JMurphyWeb** [comment]
So you used an existing email address initially?

The title of the issue is `Cannot create new user as Admin` which doesn't align with the issue that you seem to have found.

**Comment by mk4111** [comment]
Yes - I was using `ma@ka.co` as the email address to add a new user and was getting the error above. Hence I thought it was an error to do with adding a new user. 
Only when I started digging a little deeper, did I realise that that email address already belonged to someone and that there a unique constraint on the email address _only_. Does not exist for phone number.

Will update title now.

**Comment by SimonLab** [comment]
The first idea to implement a validation on the email is to check that there are not existing users with the specific email, this check is done at the moment on [the handler](https://github.com/emfoundation/ce100-app/blob/e3ea3a7f6c93397e1d36a5d6902b8fc5e5c029d8/server/handlers/people/add.js#L42). If the check fail we want to redirect to the "new user" view, so before displaying the view we need to load the data needed for the view. This means recreating the handler "user view" which is not DRY. We could require the handler inside our current handler but this start looking a bit messy.

The ideal way is to delegate the email validation to Joi.  The [extend](https://github.com/hapijs/joi/blob/v10.0.6/API.md#extendextension) API lets use add a specific new rule validation to the Join object, however at the moment the extend api only accept synchronous functions which doesn't allow us to attached an asynchronous function which get the email from our Postgres database.

So instead of having the validation on the "handler level" or the "Joi level" we can have it on the "Hapi route level". The [route configuration object](https://hapijs.com/api#route-configuration) has a property "validation" which can be a Joi object but also a function with the following signature:
```
function(value, options, next)
``` 
We can then check if the email exists already then in a callback call next.

**Comment by JMurphyWeb** [comment]
@SimonLab sounds a bit crazy, but it would be possible to put this check (ie the full DB call) into the joi validation file: server/models/admin-add-user.js

**Comment by SimonLab** [comment]
@JMurphyWeb It would be great to add the DB call in the joi validation object, we could almost do this with the extend api provide by joi but unfortunately this only works for synchronous functions at the moment. The DB call is an asynchronous function as it need to execute an SQL query on Postgres which take a bit of time. But we will update the validation as soon as Joi gives the possbility to plug asynchronous functions validation.

**Comment by JMurphyWeb** [comment]
@SimonLab given that we can use a function in Joi validation (eg [here](https://github.com/emfoundation/ce100-app/blob/master/server/models/admin-edit-org.js)), what is preventing this being async? Isn't the `next` a kind of callback?

**Comment by SimonLab** [comment]
Ok, doing the "unique email" validation inside this function will work, I just didn't see what you refered with "Joi validation".
There is a way to include the validation directly inside the [schema](https://github.com/emfoundation/ce100-app/blob/master/server/models/admin-edit-org.js#L11) with [extend](https://github.com/hapijs/joi/blob/v10.0.6/API.md#extendextension) but it has to be done with a synchronous function

**Comment by JMurphyWeb** [comment]
@SimonLab so to confirm, we can just put this into the server/model/ joi validation function that i linked to above? and call `next` when we have a response from the db? If this is right, why are you wondering about Joi's extend? Is it more suitable in some way?

**Comment by SimonLab** [comment]
Yes I'm currently adding the validation in the function.
The extend api from Joi allow to attached directly a validation function at a "Joi level" so all the validation responsibilities are delegated and encapsulated inside Joi itself. Our current solution is more at an "Hapi level" and Hapi is responsible to check that the Joi schema match our payload **and** that the email is unique.  If we could have include the email validation inside Joi we would just need to ask for the Joi schema validation.
It's just a small detail at the moment but it is interesting to see that Joi allow customised validation function (unfortunately just synchronised function at the moment) to be included in the Joi schema

**Comment by SimonLab** [comment]
Unfortunately it seems difficult to access the server methods from the validate function of the route. This means that we can't call the function server.methods.pg.people.GetBy('email',...) which allows to know if a user already exists given a specific email.

```
function(value, options, next) {
 // options doesn't have a reference to server or request object
}
````

I'm now looking if we can use [pre handler](https://hapijs.com/api#route-options) to have the email validation before calling the main handler

**Comment by JMurphyWeb** [comment]
@SimonLab where are we with this? Think we can find as simple a solution as possible as it is not a big problem just yet

**Comment by SimonLab** [comment]
This issue needs a bit of more work to have a nice integration with Hapi and Joi. I've tested multiple solution and I just need to finish testing other solutions.

**Comment by JMurphyWeb** [comment]
Looking forward, a solution that would be most helpful would be to create a helper that that returns an error similar to what the current Joi error helper does, as it would be useful to be able to pass non validation error messages back to the user.
